src/Storages/NATS/NATSLibUVAdapter.h

```diff
@@ -0,0 +1,112 @@
+#pragma once
```

Comment:
Why not just use the `src/adapter/libuv.h` header from the third party lib of nats-io (inside the contrib folder)?

Reply:
Okay now I have a better understanding and I am not sure about this:
> Probably we have calling uv_close((uv_handle_t*) nle->handle, finalCloseCb); after handle destroying and handle already assigned to another object.

I mean yes, this is what happens, but I think we are closing the handle twice, otherwise it couldn't be reallocated to another object as the handle is freed in `finalCloseCb`. Therefore I think the issue is a "double free" one. I will try to reproduce this locally and add some extra logs to see where does this happen.

P.S.: I see that the closing/destroying the NATS connections are now done with a mutex. Maybe that already solved this issue.

Reply:
> I see that the closing/destroying the NATS connections are now done with a mutex. Maybe that already solved this issue.

I tried using the latest version of the code with the original adapter and thread sanitizer, the race condition comes back.

It seems to me that nats closes the handle twice, first in natsSock_Close and second in the libuv callback:

```
E Location is file descriptor 45 destroyed by thread T2 at: 
E #0 close <null> (clickhouse+0x76a7a25) (BuildId: 7cae8b40a552b0c744d34ab6cc29dcbad660d3df) 
E #1 _closeFd build_docker/./contrib/nats-io/src/comsock.c:39:9 (clickhouse+0x1e830cdc) (BuildId: 7cae8b40a552b0c744d34ab6cc29dcbad660d3df) 
E #2 natsSock_Close build_docker/./contrib/nats-io/src/comsock.c:45:5 (clickhouse+0x1e830cdc) 
E #3 _close build_docker/./contrib/nats-io/src/conn.c:2507:13 (clickhouse+0x1e82a263) (BuildId: 7cae8b40a552b0c744d34ab6cc29dcbad660d3df) 
E #4 natsConn_close build_docker/./contrib/nats-io/src/conn.c:3959:5 (clickhouse+0x1e82c271) (BuildId: 7cae8b40a552b0c744d34ab6cc29dcbad660d3df) 
E #5 natsConnection_Close build_docker/./contrib/nats-io/src/conn.c:3977:9 (clickhouse+0x1e82c271) E #6 DB::NATSConnection::disconnectImpl() build_docker/./src/Storages/NATS/NATSConnection.cpp:162:5 (clickhouse+0x19b5f238) (BuildId: 7cae8b40a552b0c744d34ab6cc29dcbad660d3df) 
E #7 DB::NATSConnection::disconnect() build_docker/./src/Storages/NATS/NATSConnection.cpp:121:5 (clickhouse+0x19b5f289) (BuildId: 7cae8b40a552b0c744d34ab6cc29dcbad660d3df) 
E #8 DB::StorageNATS::shutdown(bool) build_docker/./src/Storages/NATS/StorageNATS.cpp:455:31 (clickhouse+0x19b4d31a) (BuildId: 7cae8b40a552b0c744d34ab6cc29dcbad660d3df) 
E #9 DB::IStorage::flushAndShutdown(bool) build_docker/./src/Storages/IStorage.h:597:9 (clickhouse+0x17d2e863) (BuildId: 7cae8b40a552b0c744d34ab6cc29dcbad660d3df)
```

```
E #0 epoll_ctl <null> (clickhouse+0x76a8af6) (BuildId: 7cae8b40a552b0c744d34ab6cc29dcbad660d3df) 
E #1 uv__platform_invalidate_fd build_docker/./contrib/libuv/src/unix/linux.c (clickhouse+0x209ff939) (BuildId: 7cae8b40a552b0c744d34ab6cc29dcbad660d3df) 
E #2 uv__poll_stop build_docker/./contrib/libuv/src/unix/poll.c:108:3 (clickhouse+0x209f940b) (BuildId: 7cae8b40a552b0c744d34ab6cc29dcbad660d3df) 
E #3 uv__poll_close build_docker/./contrib/libuv/src/unix/poll.c:159:3 (clickhouse+0x209f940b) 
E #4 uv_close build_docker/./contrib/libuv/src/unix/core.c:208:5 (clickhouse+0x209f1593) (BuildId: 7cae8b40a552b0c744d34ab6cc29dcbad660d3df) 
E #5 closeSchedulerCb(uv_handle_s*) build_docker/./contrib/nats-io/src/adapters/libuv.h:257:5 (clickhouse+0x19b5bf55) (BuildId: 7cae8b40a552b0c744d34ab6cc29dcbad660d3df) 
E #6 uv__finish_close build_docker/./contrib/libuv/src/unix/core.c:351:5 (clickhouse+0x209f256a) (BuildId: 7cae8b40a552b0c744d34ab6cc29dcbad660d3df) 
E #7 uv__run_closing_handles build_docker/./contrib/libuv/src/unix/core.c:365:5 (clickhouse+0x209f256a) 
E #8 uv_run build_docker/./contrib/libuv/src/unix/core.c:463:5 (clickhouse+0x209f256a)
```

Reply:
I found the reason, it's a mistake in NATS!!!
Thank you very much for your help, you helped me a lot

Reply:
create [issue](https://github.com/nats-io/nats.c/issues/814) in NATS project

Reply:
Wow, nicely done :clap:

Reply:
@antaljanosbenjamin the bug fixed in branch fix_814 of the nats library repository, but authors ask us to run tests before merging it. Could you create a branch fix_814 in the repository [ClickHouse/nats.c](https://github.com/ClickHouse/nats.c) so that I can make a PR in it?